[connector/routing] Use inferred contexts

andrewvc opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR allows users to use OTTL context inference in the routing connector. This makes routing table entries like the one below functional:

```
- statement: route() where span.attributes["http.method"] == "GET"
  pipelines: [traces/http]
```

Additionally, this PR fixes a minor bug where duplicate routing table entries were handled in the opposite manner as the error message indicated. The error message indicated that that the duplicate would be ignored, but in fact the original entry was replaced. I'm glad to extract this fix to a separate PR if that is preferred by reviewers.

#### Link to tracking issue

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38080

#### Testing

You can test the new routing features manually by downloading [config.yaml](https://github.com/user-attachments/files/23964620/config.yaml) which showcases the different features of this PR.

Then, run: `make otelcontribcol && bin/otelcontribcol_darwin_arm64 --config config.yaml`

To send test input:

```bash
# Routes to traces/demo: unqualified path defaults to resource context
telemetrygen traces --otlp-endpoint localhost:4317 --otlp-insecure --service demo-service --traces 1

# Routes to traces/prod: explicit resource context with qualified path
telemetrygen traces --otlp-endpoint localhost:4317 --otlp-insecure --otlp-attributes 'env="production"' --traces 1

# Routes to traces/default: no matching conditions
telemetrygen traces --otlp-endpoint localhost:4317 --otlp-insecure --service other-service --traces 1

# Routes to traces/http: span context inferred from qualified path
telemetrygen traces --otlp-endpoint localhost:4317 --otlp-insecure --telemetry-attributes 'http.method="GET"' --traces 1
```

#### Documentation

New docs have been added to the README

andrewvc committed:
Checkpoint

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Merge remote-tracking branch 'origin' into 38080-connector-routing-context

andrewvc committed:
Improvements and simplification

andrewvc committed:
Test improvements, cleanup

andrewvc committed:
More cleanup

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Fix statement continue bug

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Remove unneeded config.yaml

andrewvc committed:
Make linter happy

andrewvc committed:
Add chlogs

andrewvc committed:
Cleanup inference test

andrewvc committed:
Better test naming

andrewvc committed:
Cleanup / enhance signal specific tests

andrewvc committed:
Cleanup README notes

andrewvc committed:
Add clarifying comments

andrewvc committed:
Clarify normalization

andrewvc committed:
More clarity in comments

andrewvc committed:
Fix formatting

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc commented on connector/routingconnector/router.go:
```diff
@@ -81,80 +80,121 @@ type routingItem[C any] struct {
+	// exist in resource anyway, and non-ambiguous paths (like "body", "severity_text", "name")
+	// only exist in one context regardless of priority. That said, the order is sorted based on which
+	// events are most common in practice, hence 'span' is first.
+	priorities := []string{
```
@edmocosta do you have any advice for how to handle potential new types that get added to OTel in the future? I'm not nuts about this list sitting here, but it's no different than the previous case statements I suppose.

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Merge branch '38080-connector-routing-context' of github.com:andrewvc/opentelemetry-collector-contrib into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Merge remote-tracking branch 'andrewvc/38080-connector-routing-context' into 38080-connector-routing-context

andrewvc committed:
Improved README

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta commented on pkg/ottl/parser_collection.go:
```diff
@@ -428,6 +411,34 @@ func WithContextInferenceConditions(conditions []string) ParserCollectionContext
+// compatibility when ambiguous paths (e.g., 'attributes') should resolve to a specific context.
+//
+// Experimental: *NOTE* this API is subject to change or removal in the future.
+func WithContextInferrerPriorities[R any](priorities []string) ParserCollectionOption[R] {
```
We don't need to expose this functionality, please see my other comment related to context inference priorities for reference.

edmocosta reviewed (commented):
Hi @andrewvc thanks for working on this. I've left a few comments. I didn't fully tested it yet, and I'm planning to do it after we address them.

Regarding the `request` context, do we have any plans to properly integrate it with the context inferrer? I'm not sure if that's really possible at the moment, and one idea would be waiting for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41879 and deprecate the existing `request` impl for inferred configurations.

andrewvc committed:
Remove unnecessary custom prioritization

andrewvc committed:
PR Feedback, rename inference_test to router and add logging

andrewvc committed:
First stab at restoring prependPathsContext

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Restore external interface

andrewvc committed:
Merge remote-tracking branch 'andrewvc/38080-connector-routing-context' into 38080-connector-routing-context

andrewvc committed:
Rename WithContextInferenceDefaultContext and add unit tests

andrewvc commented on pkg/ottl/parser_collection.go:
I've restored the file except my one added function. I kept the improved IMHO comments if that's OK.

andrewvc committed:
Restore old test defaults that had no impact to minimize diff

andrewvc committed:
Restore accidental transform changes

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

Conflict resolutions:

connector/routingconnector/router.go:
- Kept HEAD's context inference approach (parserCollection +
  WithContextInferenceDefaultContext) and corrected duplicate handling
  (dupeFound + continue).
- Added route.action = item.Action from origin/main so the action:
  copy/move feature introduced in main works correctly.
- Simplified `if !dupeFound { r.routeSlice = append(...) }` to a
  plain append since dupeFound always triggers a continue before
  reaching that line.

connector/routingconnector/README.md:
- Kept HEAD's improved descriptions for table.condition,
  table.statement, and table.context (including context inference docs).
- Added origin/main's new table.action documentation.

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
[connector/routing] Add otelcol.client/grpc.metadata routing and deprecate request context

- Emit deprecation Warn when context: "request" is used, directing users to
  otelcol.client.metadata or otelcol.grpc.metadata OTTL paths instead
- Add integration tests for otelcol.client.metadata and otelcol.grpc.metadata
  routing in TestLogsConnectorDetailed, including a behavioral test documenting
  that index-based access [0] only checks a single value unlike legacy request
  context which checks all values
- Add TestOtelcolPathsParseWithExplicitContext confirming otelcol.* paths parse
  correctly with explicit context (inference unsupported for otelcol.* paths)
- Add TestRequestContextCannotBeInferred and TestRequestContextDeprecationWarning
- Update README with deprecation notice, context table, and updated examples

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -82,80 +81,99 @@ type routingItem[C any] struct {
+	// like the transform processor. The OTTL context inferrer needs access to all context
+	// parsers to properly determine which context to use based on paths, functions, and enums.
+	// The one-time initialization cost is minimal compared to the complexity and fragility
+	// of trying to pre-determine which contexts are needed via statement inspection.
```
```suggestion
```

We don't need to explain this change reasoning on the code. Once it's merged, the difference from this and the previous version fades away, and this comment doesn't provide any extra information that can't be understood by reading the code itself.

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -82,80 +81,99 @@ type routingItem[C any] struct {
+//
+// The length check is technically redundant since registerRouteConsumers always passes exactly
+// one statement to the parser, and the OTTL parser produces one parsed statement per input.
+// However, it serves as defense-in-depth against future bugs in either this code or the OTTL library.
```
```suggestion

```

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,64 +223,74 @@ func (r *router[C]) normalizeConditions() {
 			exporters := strings.Join(pipelineNames, ", ")
 			r.logger.Warn(fmt.Sprintf(`Statement %q already exists in the routing table, the route with target pipeline(s) %q will be ignored.`, item.Statement, exporters))
+			// Without this continue, the duplicate's pipelines would overwrite the original
+			// route's consumer, contradicting the warning message above.
```
```suggestion
```

It's explaining the bug-fix, not really necessary IMO.

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,64 +223,74 @@ func (r *router[C]) normalizeConditions() {
+
+		route.statementContext = item.Context
+		if item.Context == "request" {
+			r.logger.Warn("The 'request' context is deprecated. Use 'otelcol.client.metadata[\"key\"][0]' "+
```
```suggestion
			r.logger.Warn("The 'request' context is deprecated. Use 'otelcol.client.metadata[\"key\"]' "+
```

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,64 +223,74 @@ func (r *router[C]) normalizeConditions() {
+		route.statementContext = item.Context
+		if item.Context == "request" {
+			r.logger.Warn("The 'request' context is deprecated. Use 'otelcol.client.metadata[\"key\"][0]' "+
+				"(HTTP/client metadata) or 'otelcol.grpc.metadata[\"key\"][0]' (gRPC metadata) instead.",
```
```suggestion
				"(HTTP/client metadata) or 'otelcol.grpc.metadata[\"key\"]' (gRPC metadata) instead.",
```

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,64 +223,74 @@ func (r *router[C]) normalizeConditions() {
+			if item.Context == "" {
+				// Context is empty, try to infer it
+				// Default to resource context if inference fails or ambiguous (though priorities handle ambiguity)
+				result, err = r.parserCollection.ParseStatements(statementsGetter, ottl.WithContextInferenceDefaultContext(ottlresource.ContextName))
```
I think we shouldn't change the parser collection to support this specific use-case. `ParseStatements` should be used to parse inferred contexts only (prefixed paths), and IMO it shouldn't silently rewrite the statements's based on some default context option. My suggestion here would be introducing a breaking change to the router connector, and treat empty contexts as inferred ones, so we don't need to support this deprecate pattern longer. WDYT @TylerHelmuth @evan-bradley ?

TylerHelmuth replied:
If we can add this feature to the ParserCollection construction instead of the ParseStatements then I think this feature makes sense. To add it to ParseStatements feels like an override of the originally created ParserCollection we don't want to do.

TylerHelmuth replied:
Today if a ParseCollection is given an OTTL statement with no context prefixes I forget how it determines which context to use. Doesn't the value have to get passed in explicitly? Couldn't we explicitly pass in `resource`?

TylerHelmuth replied:
Oh I see now what you mean @edmocosta, a default context makes no sense in the future where contextless paths are not allowed via the grammar. So I agree we dont want a default context concept anywhere.

andrewvc replied:
I've resolved all the other review comments, just let me know which direction you'd like to take this section of code @edmocosta and @TylerHelmuth

edmocosta replied:
Thanks @andrewvc, we've discussed it and the approach we'd like to take is:
 - Rollback all parse collection changes as we don't really want to introduce a default context concept (comments, docs and non-related changes shouldn't be on this PR as well).
 - Change the `registerRouteConsumers` to retry the statement parsing as https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44762/changes#r3247241272, so it supports new and old formats, with proper logging. 

This is a transition approach, after some time with the warning message in place, we might break the config compatibility to enforce the context prefixes to all statements.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,60 @@ If you are not already familiar with connectors, you may find it helpful to firs
-- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context.
+- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).
+- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. May not be used for `request` context. Generally `condition` is preferred since it is more terse.
+- `table.context (optional)`: the [OTTL Context] in which the condition/statement will be evaluated. Only `resource`, `span`, `metric`, `datapoint`, `log`, and `request` are supported. In most cases, you should omit this field and use context-qualified paths instead, which allows the context to be inferred automatically. If specified, takes precedence over inference. Defaults to `resource` when n ...
```
I think we should call out the `request` "context" deprecation here.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,60 @@ If you are not already familiar with connectors, you may find it helpful to firs
-- `table.context (optional, default: resource)`: the [OTTL Context] in which the statement will be evaluated. Currently, only `resource`, `span`, `metric`, `datapoint`, `log`, and `request` are supported.
-- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. May not be used for `request` context.
-- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context.
+- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).
```
Let's ask users to use qualified paths here, similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/48330/changes#diff-6a46f242e342c701ac32dc243603396800161de24c34ab4faff72c665216190fR40.

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -82,80 +81,99 @@ type routingItem[C any] struct {
-			r.logParser = parser
-		} else {
-			errs = errors.Join(errs, err)
+
```
`otelcol` can be used as an independently context, and we should register/handle it on this connector as well, so conditions that only uses `otelcol` paths would work properly.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,60 @@ If you are not already familiar with connectors, you may find it helpful to firs
+| Log | `log.` | `log.body`, `log.attributes["level"]` |
+| Metric | `metric.` | `metric.name` |
+| Datapoint | `datapoint.` | `datapoint.attributes["host"]` |
+| Any (requires explicit context) | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |
```
```suggestion
| Otelcol | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,60 @@ If you are not already familiar with connectors, you may find it helpful to firs
+| Datapoint | `datapoint.` | `datapoint.attributes["host"]` |
+| Any (requires explicit context) | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |
+
+The `otelcol.client.metadata` and `otelcol.grpc.metadata` paths provide access to incoming HTTP and gRPC request metadata respectively, and are valid in all signal contexts. **Important:** Context inference does not work for `otelcol.*` paths — you must set the `context` field explicitly (e.g., `context: resource`). Note that gRPC lowercases all metadata keys.
```
```suggestion
The `otelcol.client.metadata` and `otelcol.grpc.metadata` paths provide access to incoming HTTP and gRPC request metadata respectively, and are valid in all signal contexts.
```
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44762/changes#r3233998184

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,60 @@ If you are not already familiar with connectors, you may find it helpful to firs
+
+The `otelcol.client.metadata` and `otelcol.grpc.metadata` paths provide access to incoming HTTP and gRPC request metadata respectively, and are valid in all signal contexts. **Important:** Context inference does not work for `otelcol.*` paths — you must set the `context` field explicitly (e.g., `context: resource`). Note that gRPC lowercases all metadata keys.
+
+**Using explicit context:** If you prefer, you can still set the `context` field explicitly. This can be useful when working with unqualified paths:
```
```suggestion
```

Let's not suggest this as the goal is to deprecate contextless paths.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -64,7 +105,50 @@ The full list of settings exposed for this connector are documented in [config.g
 
 ## Examples
```
All examples should be updated/use qualified paths and inferred contexts, as it's the format we want to guide users to. We can add a note similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/README.md?plain=1#L21 linking to the previous docs as well, and maybe a single example where setting the context manually makes sense, such as the examples listed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#context-inference).

andrewvc committed:
[connector/routing] Fix otelcol context registration and evaluation

- Register otelcol parser under ottlotelcol.ContextName ("otelcol") instead
  of ottlresource.ContextName ("resource"), which caused all tests to fail
- Add missing error check after otelcolParser creation
- Add otelcolStatement field to routingItem and handle
  *ottl.Statement[*ottlotelcol.TransformContext] in the type switch
- Add "otelcol" case to traces, logs, and metrics consumers so that
  otelcol-context routes are evaluated and routed correctly

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
Merge branch '38080-connector-routing-context' of github.com:andrewvc/opentelemetry-collector-contrib into 38080-connector-routing-context

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Improve README based on feedback

andrewvc committed:
Merge remote-tracking branch 'andrewvc/38080-connector-routing-context' into 38080-connector-routing-context

andrewvc committed:
[connector/routing] Update README examples to use context-qualified paths

- Add > [!NOTE] callout at the top of Examples pointing users to the
  Context Inference section, matching the style of the transform processor
- Update all examples to use context-qualified paths (e.g.,
  resource.attributes["x"], log.severity_number) with inferred context,
  removing explicit context: fields except where required
- Rename and reframe the tenant example as the canonical case where
  explicit context is required (otelcol.* paths cannot be inferred)
- Add explanation for the archive example where condition: "true" requires
  explicit context because there are no paths to infer from
- Remove the "Alternative: Explicit context with unqualified paths"
  snippet and backward compatibility note from the Context Inference
  section, as the goal is to guide users toward qualified paths
- Update match_once migration "after" examples to use qualified paths

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
[connector/routing] Clean up remaining README review feedback

- Fix typos in table.condition description
- Reword table.statement to not reference deprecated request context
- Move request context deprecation to a prominent inline note in
  table.context, with a link to Limitations, rather than burying it
  in a parenthetical
- Remove [0] index from otelcol.client/grpc.metadata path references
  in Limitations section per reviewer suggestion

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,64 +227,75 @@ func (r *router[C]) normalizeConditions() {
+		if item.Context == "request" {
+			r.logger.Warn("The 'request' context is deprecated. Use 'otelcol.client.metadata[\"key\"]' "+
+				"(HTTP/client metadata) or 'otelcol.grpc.metadata[\"key\"]' (gRPC metadata) instead.",
+				zap.String("condition", item.Condition))
+			route.requestCondition, err = parseRequestCondition(item.Condition)
+			if err != nil {
+				return err
+			}
+		} else {
+			statementsGetter := ottl.NewStatementsGetter([]string{item.Statement})
+			var result any
+			if item.Context == "" {
+				// Context is empty, try to infer it
+				// Default to resource context if inference fails or ambiguous (though priorities handle ambiguity)
+				result, err = r.parserCollection.ParseStatements(statementsGetter, ottl.WithContextInferenceDefaultContext(ottlresource.ContextName))
+			} else {
+				// Context is explicit
+				result, err = r.parserCollection.ParseStatementsWithContext(item.Context, statementsGetter, true)
+			}
```
```suggestion
			if item.Context == "" {
				// Parse using the inferred context. If that fails, retry with the "resource"
				// context so older configs still work: an empty context used to imply "resource".
				result, err = r.parserCollection.ParseStatements(statementsGetter)
				if err != nil {
					result, err = r.parserCollection.ParseStatementsWithContext(ottlresource.ContextName, statementsGetter, true)
				}
			} else {
				result, err = r.parserCollection.ParseStatementsWithContext(item.Context, statementsGetter, true)
			}
```

This should support older and new usages, given preference to the context inference. When the retry with a fixed `resource` context succeeds, users will see a log message asking them to add the context prefixes to their paths, which is desired. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44762#discussion_r3247288991 for reference.

edmocosta commented on connector/routingconnector/metrics.go:
```diff
@@ -77,6 +78,20 @@ func (c *metricsConnector) ConsumeMetrics(ctx context.Context, md pmetric.Metric
 			}
+		case "otelcol":
+			otx := ottlotelcol.NewTransformContextPtr()
+			defer otx.Close()
+			_, isMatch, err := route.otelcolStatement.Execute(ctx, otx)
```
```suggestion
			otx := ottlotelcol.NewTransformContextPtr()
			_, isMatch, err := route.otelcolStatement.Execute(ctx, otx)
			otx.Close()
```
To not call `defer` in the for loop.

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -275,9 +308,7 @@ func key(entry RoutingTableItem) string {
 		return entry.Statement
 	case "request":
 		return "[request] " + entry.Condition
+	default:
```
Now, when the context is inferred, `item.Context == ""` does not necessary means `resource` anymore.  `entry.Context` is likely to be empty, matching the first case condition. Considering `entry.Statement` does not contains the rewritten statement with context paths, we might have conflicts on this key algorithm that would lead to false positive/negative duplicates. E.g:

```yaml
table:
  # key = `[span] span.attributes["foo"] != "bar"`
  - context: span
    condition: span.attributes["foo"] != "bar"  

  # key = `span.attributes["foo"] != "bar"` (inferred context is also span)
  - condition: span.attributes["foo"] != "bar"
```

andrewvc replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44762/changes/96d123c76810cf32d80557d9afcc22adf2b2dc72

edmocosta replied:
This issue wasn't fixed, for example, the following isn't detected as duplicate, and under the hood they're the same statement. I'm not sure if we can fix it at the routing connector level without exposing the re-written statement with paths. Let me double check it. 

```yaml
connectors:
  routing:
    table:
      # key = `[span] attributes["foo"] != "bar"`
      - context: span
        condition: attributes["foo"] != "bar"  

      # key = `[span] span.attributes["foo"] != "bar"`
      - condition: span.attributes["foo"] != "bar"
```

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -82,80 +83,101 @@ type routingItem[C any] struct {
-			buildLog = true
-		}
+func (r *router[C]) buildParsers(_ []RoutingTableItem, settings component.TelemetrySettings) error {
+	otelcolParser, err := ottlotelcol.NewParser(
```
We need to change the config [validation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/418d2d26734ae4c3359a1162b798ee4bd6c6c2cb/connector/routingconnector/config.go#L102) to also support it. It seems we're also missing unit tests for that.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,49 @@ If you are not already familiar with connectors, you may find it helpful to firs
-- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context.
+- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).
+- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. Generally `condition` is preferred since it is more terse.
+- `table.context (optional)`: the [OTTL Context] in which the condition/statement will be evaluated. Supported values: `resource`, `span`, `metric`, `datapoint`, `log`. **Deprecated:** `request` — use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths instead (see [Limitations](#limitations)). In most cases this field should be omitted; the context is inferred automatically from context-q ...
```
```suggestion
- `table.context (optional)`: the [OTTL Context](#supported-contexts) in which the condition/statement will be evaluated. **Deprecated:** `request` — use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths instead (see [Limitations](#limitations)). In most cases this field should be omitted; the context is inferred automatically from context-qualified paths. If specified, it takes precedence over inference.
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,49 @@ If you are not already familiar with connectors, you may find it helpful to firs
-- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. May not be used for `request` context.
-- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context.
+- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).
+- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. Generally `condition` is preferred since it is more terse.
```
```suggestion
- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. Generally `condition` is preferred since it is more terse. May not be used with the deprecated `request` context.
```

It's still valid.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,49 @@ If you are not already familiar with connectors, you may find it helpful to firs
+
+This approach makes it immediately clear which attributes you're accessing without needing a separate `context` field.
+
+**Supported context-qualified paths:**
```
```suggestion
### Supported contexts
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -34,19 +34,49 @@ If you are not already familiar with connectors, you may find it helpful to firs
+| Log | `log.` | `log.body`, `log.attributes["level"]` |
+| Metric | `metric.` | `metric.name` |
+| Datapoint | `datapoint.` | `datapoint.attributes["host"]` |
+| Otelcol | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |
```
```suggestion
| Context     | Path prefix  | Example                                                                          |
|-------------|--------------|----------------------------------------------------------------------------------|
| [Resource]  | `resource.`  | `resource.attributes["service.name"]`                                            |
| [Span]      | `span.`      | `span.attributes["http.method"]`                                                 |
| [Log]       | `log.`       | `log.body`, `log.attributes["level"]`                                            |
| [Metric]    | `metric.`    | `metric.name`                                                                    |
| [Datapoint] | `datapoint.` | `datapoint.attributes["host"]`                                                   |
| [OtelCol]   | `otelcol.`   | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |

[resource]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlresource/README.md
[span]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlspan/README.md
[metric]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlmetric/README.md
[datapoint]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottldatapoint/README.md
[log]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottllog/README.md
[otelcol]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlotelcol/README.md
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -64,7 +94,51 @@ The full list of settings exposed for this connector are documented in [config.g
+
+### Route logs based on tenant (explicit context required)
+
+When conditions use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths, the `context` field must be set explicitly because these paths cross-cut all signals and cannot be inferred. This is the primary case where an explicit `context` is needed.
```
```suggestion
### Route logs based on tenant

```

We actually can use `otelcol` statements without explicitly set the `context` key.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -82,11 +156,13 @@ connectors:
-        condition: request["X-Tenant"] == "ecorp"
+      - context: resource
+        condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
         pipelines: [logs/ecorp]
```
```suggestion
      - condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
        pipelines: [logs/acme]
      - condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
        pipelines: [logs/ecorp]
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -64,7 +94,51 @@ The full list of settings exposed for this connector are documented in [config.g
+> [!NOTE]
+> The examples below use context-qualified paths, which is the recommended configuration style. The explicit `context` field is still supported for backward compatibility but is no longer the primary documentation style. See [Context Inference](#context-inference) for details.
+
+### Route traces using context-qualified paths
```
If all examples are using context-qualified paths, we probably don't need to add this extra one as it doesn't show any specific case.

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -206,14 +277,13 @@ exporters:
-        condition: request["X-Tenant"] == "acme"
+      - context: resource
+        condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
         pipelines: [logs/acme]
-      - context: request
-        condition: request["X-Tenant"] == "ecorp"
+      - context: resource
+        condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
         pipelines: [logs/ecorp]
```
```suggestion
      - condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
        pipelines: [logs/acme]
      - condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
        pipelines: [logs/ecorp]
```

edmocosta commented on connector/routingconnector/README.md:
```suggestion
    - condition: resource.attributes["env"] == "prod"
      pipelines: [ logs/prod ]
    - condition: resource.attributes["env"] == "dev"
      pipelines: [ logs/dev ]
    - condition: resource.attributes["region"] == "east"
      pipelines: [ logs/east ]
    - condition: resource.attributes["region"] == "west"
      pipelines: [ logs/west ]
```

Adding `resource.` prefix.

andrewvc committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Apply suggestion from @edmocosta

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Apply suggestion from @edmocosta

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Remove traces example from README

andrewvc committed:
Cleanup routing based on tenant example

andrewvc committed:
[connector/routing] Add otelcol context to config validation

Add "otelcol" as a valid context value in Config.Validate() and add
corresponding test cases for both statement and condition forms.

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
[connector/routing] Fix import ordering in logs.go and metrics.go

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
[connector/routing] Fix duplicate route detection for inferred contexts

The key() function previously used the raw entry.Context value, so a route
with explicit context: span and a semantically identical route with inferred
span context (empty context field) would produce different keys and both be
registered. Now parsing happens before the duplicate check so the resolved
statementContext drives the key, making explicit and inferred routes
for the same statement correctly collapse.

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
[connector/routing] Remove WithContextInferenceDefaultContext; add fallback warning

Per reviewer consensus: remove the WithContextInferenceDefaultContext option
from pkg/ottl (keeps ParseStatements API clean, no default-context concept in
the library). The backward-compat resource fallback is handled entirely in the
connector: when context inference fails for an empty-context route, retry with
ParseStatementsWithContext("resource") and emit a deprecation warning directing
users to add context-qualified path prefixes (e.g. resource.attributes["x"]).

Assisted-by: Claude Sonnet 4.6

andrewvc committed:
update schema

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc commented:
@edmocosta I think this is ready for another review

edmocosta commented on connector/routingconnector/logs.go:
```diff
@@ -76,6 +77,20 @@ func (c *logsConnector) ConsumeLogs(ctx context.Context, ld plog.Logs) error {
+		case "otelcol":
+			otx := ottlotelcol.NewTransformContextPtr()
+			defer otx.Close()
+			_, isMatch, err := route.otelcolStatement.Execute(ctx, otx)
```
```suggestion
			_, isMatch, err := route.otelcolStatement.Execute(ctx, otx)
			otx.Close()
```

To not call `defer` in the `for` loop

edmocosta committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -401,7 +435,14 @@ routing:
        pipelines: [ logs/east ]
     - condition: attributes["region"] == "west"
-       pipelines: [ logs/west ]
+    - condition: resource.attributes["env"] == "prod"
```
I think this new content should replace the above conditions without context's prefix. It seems it got appended to the list instead.

andrewvc replied:
It is now removed above

edmocosta commented on connector/routingconnector/traces.go:
```diff
@@ -76,6 +77,20 @@ func (c *tracesConnector) ConsumeTraces(ctx context.Context, td ptrace.Traces) e
 			}
+		case "otelcol":
+			otx := ottlotelcol.NewTransformContextPtr()
+			defer otx.Close()
```
Let's not call defer in a for loop.

andrewvc replied:
Done!

edmocosta commented on pkg/ottl/parser_collection.go:
Please revert all parser collection changes. If needed, we can document it better later in a separate PR.

andrewvc replied:
Done!

edmocosta commented on connector/routingconnector/router.go:
```diff
@@ -205,79 +227,96 @@ func (r *router[C]) normalizeConditions() {
-					return err
+					r.logger.Warn("Routing condition uses unqualified paths which default to 'resource' context for backward compatibility. "+
+						"Update the condition to use context-qualified paths (e.g. 'resource.attributes[\"x\"]') to silence this warning.",
+						zap.String("statement", item.Statement))
```
```suggestion
```

The parser collection already emits a standardized log message about that if `ParseStatementsWithContext` rewrite the statements. When it fails to parse, the `err` is probably accurate and enough. The parsing statement might have prefixes but be invalid, making this warn message misleading.

edmocosta reviewed (commented):
Thanks @andrewvc for addressing all suggestions. I'm still reviewing/running a few tests, but it seems we're almost there! :)

andrewvc committed:
Just execute the close earlier, no defer

andrewvc committed:
Merge remote-tracking branch 'origin/main' into 38080-connector-routing-context

andrewvc committed:
Revert all parser collection changes

andrewvc committed:
Merge remote-tracking branch 'andrewvc/38080-connector-routing-context' into 38080-connector-routing-context

andrewvc committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

andrewvc committed:
Remove old style from README without prefix

andrewvc commented:
@edmocosta I've now implemented all suggested fixes FYI

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
 			expectSinkD: plog.Logs{},
 		},
+		// otelcol.client.metadata routes (HTTP client metadata; requires explicit context since
+		// otelcol.* paths cannot be inferred — use context: "resource" explicitly)
```
```suggestion
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_client_metadata/http_match",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_client_metadata/http_no_match",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_client_metadata/no_metadata",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_client_metadata/http_value_not_at_index0",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.client.metadata["X-Tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_grpc_metadata/grpc_match",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.grpc.metadata["x-tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.grpc.metadata["x-tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/logs_test.go:
```diff
@@ -887,6 +1002,84 @@ func TestLogsConnectorDetailed(t *testing.T) {
+		{
+			name: "otelcol_grpc_metadata/grpc_no_match",
+			cfg: testConfig(
+				withRoute("resource", `otelcol.grpc.metadata["x-tenant"][0] == "acme"`, idSink0),
```
```suggestion
				withRoute("otelcol", `otelcol.grpc.metadata["x-tenant"][0] == "acme"`, idSink0),
```

edmocosta commented on connector/routingconnector/README.md:
````diff
@@ -34,19 +34,56 @@ If you are not already familiar with connectors, you may find it helpful to firs
+The routing connector supports OTTL context inference, allowing you to write clearer and more maintainable routing conditions using context-qualified paths. This is the recommended approach for specifying routing conditions.
+
+```yaml
+# Recommended: Use context-qualified paths for clarity
````
```suggestion
```

edmocosta commented on connector/routingconnector/README.md:
```diff
@@ -394,14 +428,15 @@ routing:
-       pipelines: [ logs/dev ]
-    - condition: attributes["region"] == "east"
-       pipelines: [ logs/east ]
     - condition: attributes["region"] == "west"
```
```suggestion
```

edmocosta commented on connector/routingconnector/router_test.go:
```diff
@@ -0,0 +1,334 @@
+	_, err := newRouter(routeTable, nil,
+		func(...pipeline.ID) (consumer.Logs, error) { return sink, nil },
+		componenttest.NewNopTelemetrySettings())
+	require.Error(t, err)
```
```suggestion
	require.ErrorContains(t, err, `segment "request" from path "resource.request[X-Tenant]" is not a valid path`)
```

edmocosta commented on connector/routingconnector/router_test.go:
```diff
@@ -0,0 +1,334 @@
+			assert.Equal(t, "resource", router.routeSlice[0].statementContext)
+		})
+	}
+}
```
```suggestion

```

edmocosta commented on connector/routingconnector/router_test.go:
```diff
@@ -0,0 +1,334 @@
+		})
+	}
+}
+
```
```suggestion
```

edmocosta commented on connector/routingconnector/router_test.go:
```diff
@@ -0,0 +1,334 @@
+			signal:          pipeline.SignalMetrics,
+			context:         "datapoint",
+			expectedContext: "datapoint",
+		},
```
```suggestion
		},
		{
			name:            "otelcol/log signal with qualified paths infers otelcol",
			condition:       `otelcol.client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalLogs,
			context:         "",
			expectedContext: "otelcol",
		},
		{
			name:            "otelcol/log signal with explicit context overrides inference",
			condition:       `client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalLogs,
			context:         "otelcol",
			expectedContext: "otelcol",
		},
		{
			name:            "otelcol/metric signal with qualified paths infers otelcol",
			condition:       `otelcol.client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalMetrics,
			context:         "",
			expectedContext: "otelcol",
		},
		{
			name:            "otelcol/metric signal with explicit context overrides inference",
			condition:       `client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalMetrics,
			context:         "otelcol",
			expectedContext: "otelcol",
		},
		{
			name:            "otelcol/trace signal with qualified paths infers otelcol",
			condition:       `otelcol.client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalTraces,
			context:         "",
			expectedContext: "otelcol",
		},
		{
			name:            "otelcol/trace signal with explicit context overrides inference",
			condition:       `client.metadata["foo"][0] == "bar"`,
			signal:          pipeline.SignalTraces,
			context:         "otelcol",
			expectedContext: "otelcol",
		},
```

edmocosta committed:
Apply suggestions from code review

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

edmocosta committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

andrewvc committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta committed:
Fix route deduplication key conflicts

edmocosta committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta committed:
Fix misleading error message when context inference fails

edmocosta committed:
Merge branch '38080-connector-routing-context' of https://github.com/andrewvc/opentelemetry-collector-contrib into 38080-connector-routing-context

TylerHelmuth commented on connector/routingconnector/router.go:
```diff
@@ -205,79 +228,101 @@ func (r *router[C]) normalizeConditions() {
-					return err
+					r.logger.Debug("Failed to parse statement with context inference, retrying with 'resource' context.", zap.Error(err))
+					var retryErr error
+					result, retryErr = r.parserCollection.ParseStatementsWithContext(ottlresource.ContextName, statementsGetter, true)
```
I'm glad we were able to implement `resource` as the fallback

mwear reviewed (approved):
Mostly LGTM. One non-blocking question: Should we have a changelog entry for the request context deprecation?

edmocosta committed:
Add "request" deprecation change log

edmocosta committed:
Merge branch 'main' into 38080-connector-routing-context

edmocosta commented:
> Mostly LGTM. One non-blocking question: Should we have a changelog entry for the request context deprecation?

Done! thank you, @mwear.
